File: Week_03/kotlin/src/_105_ConstructBinaryTreeFromPreorderAndInorderTraversal.kt
class _105_ {
    fun buildTree(preorder: IntArray, inorder: IntArray): TreeNode? {
        val map = HashMap<Int, Int>()
        inorder.forEachIndexed { idx, v -> map[v] = idx }
        fun dfs(pStart: Int, pEnd: Int, iStart: Int, iEnd: Int): TreeNode? {
            if (pStart > pEnd) return null
            val root = TreeNode(preorder[pStart])
            val idx = map[preorder[pStart]]!!
            val len = idx - iStart
            root.left = dfs(pStart + 1, pStart + len, iStart, idx - 1)
            root.right = dfs(pStart + 1 + len, pEnd, idx + 1, iEnd)
            return root
        }
        return dfs(0, preorder.lastIndex, 0, inorder.lastIndex)
    }
}

class TreeNode(var `val`: Int) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}

File: Week_03/cpp/src/_105_ConstructBinaryTreeFromPreorderAndInorderTraversal.cpp
#include <vector>
#include <unordered_map>

using namespace std;

class Solution
{
public:
    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder)
    {
        unordered_map<int, int> table;
        for (int i = 0; i < inorder.size(); i++)
        {
            table[inorder[i]] = i;
        }

        return dfs(preorder, 0, preorder.size() - 1, inorder, 0, inorder.size() - 1, table);
    }
    TreeNode *dfs(vector<int> &preorder, int pStart, int pEnd, vector<int> &inorder, int iStart, int iEnd, unordered_map<int, int> &table)
    {
        if (pStart > pEnd)
            return NULL;
        auto root = new TreeNode(preorder[pStart]);
        int idx = table[preorder[pStart]];
        int len = idx - iStart;
        root->left = dfs(preorder, pStart + 1, pStart + len, inorder, iStart, idx - 1, table);
        root->right = dfs(preorder, pStart + 1 + len, pEnd, inorder, idx + 1, iEnd, table);
        return root;
    }
};

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};